hashicorp/dev-portal
# [wip] feat: ThemedImage

## 🔗 Relevant links



- [Preview link](https://dev-portal-git-kscreates-themed-image-remark-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1204334498378837) 🎟️

## 🗒️ What

|  dark  | light  |  
|---|---|
|  <img width="979" alt="Screenshot 2023-06-23 at 10 45 40 AM" src="https://github.com/hashicorp/dev-portal/assets/36613477/9a852a2e-0c6d-4ed8-b961-65d40f544c5a"> |  <img width="983" alt="Screenshot 2023-06-23 at 10 45 47 AM" src="https://github.com/hashicorp/dev-portal/assets/36613477/e1274726-5553-45b4-a2fb-2dc8b4bedfd1"> |   



### MDX Component
This PR adds support for a `MdxThemedImage` in tutorials content. This component accepts a srcset for both dark and light assets, which render conditionally based on theme state. 

**Example authoring interface**
```
<ThemedImage
  src={{
    dark: '/img/themed/dark-test.png',
    light: '/img/themed/light-test.png'
  }}
  alt="testing"
/>
```

### Remark Plugin
To use `next/image`, we need to pass the assets width / height. Authors can pass this directly via props. However, if they don't this PR also adds a new remark plugin to inject the width / height onto the component, if able to calculate. 

The plugin also rewrites the assets `src`, since the files are hosted in the tutorials repo and served via the `mktg-content-api` (a la GH API). This functions similar to `rewriteStaticAssets` plugin already used for plain images on tutorials and docs content. 

## 🧪 Testing

- [Visit the preview](https://dev-portal-git-kscreates-themed-image-remark-hashicorp.vercel.app/well-architected-framework/operational-excellence/operational-excellence-introduction) where `ThemedImage` is in use via the staging API. 
- Switch themes, validate that the asset changes
- Check the markup, validate that lazy loaded `img` tags (via `next/image`) are being used

## 💭 Anything else?

I've tested this locally with the tutorials repo workflow and it works. Note that this preview's src url path will be the same as prod (from mktg-content-api).
